<!-- This comment will put IE 6, 7 and 8 in quirks mode -->
<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">
<html xmlns="http://www.w3.org/1999/xhtml">
<head>
<meta http-equiv="Content-Type" content="text/xhtml;charset=UTF-8"/>
<title>T-Coffee: fastal.c File Reference</title>
<link href="tabs.css" rel="stylesheet" type="text/css"/>
<link href="search/search.css" rel="stylesheet" type="text/css"/>
<script type="text/javaScript" src="search/search.js"></script>
<link href="doxygen.css" rel="stylesheet" type="text/css"/>
</head>
<body onload='searchBox.OnSelectItem(0);'>
<!-- Generated by Doxygen 1.6.1 -->
<script type="text/javascript"><!--
var searchBox = new SearchBox("searchBox", "search",false,'Search');
--></script>
<div class="navigation" id="top">
  <div class="tabs">
    <ul>
      <li><a href="index.html"><span>Main&nbsp;Page</span></a></li>
      <li><a href="pages.html"><span>Related&nbsp;Pages</span></a></li>
      <li><a href="annotated.html"><span>Classes</span></a></li>
      <li class="current"><a href="files.html"><span>Files</span></a></li>
      <li>
        <div id="MSearchBox" class="MSearchBoxInactive">
        <img id="MSearchSelect" src="search/search.png"
             onmouseover="return searchBox.OnSearchSelectShow()"
             onmouseout="return searchBox.OnSearchSelectHide()"
             alt=""/>
        <input type="text" id="MSearchField" value="Search" accesskey="S"
             onfocus="searchBox.OnSearchFieldFocus(true)" 
             onblur="searchBox.OnSearchFieldFocus(false)" 
             onkeyup="searchBox.OnSearchFieldChange(event)"/>
        <a id="MSearchClose" href="javascript:searchBox.CloseResultsWindow()"><img id="MSearchCloseImg" border="0" src="search/close.png" alt=""/></a>
        </div>
      </li>
    </ul>
  </div>
  <div class="tabs">
    <ul>
      <li><a href="files.html"><span>File&nbsp;List</span></a></li>
      <li><a href="globals.html"><span>File&nbsp;Members</span></a></li>
    </ul>
  </div>
</div>
<div class="contents">
<h1>fastal.c File Reference</h1>
<p>Source code for the fastal algorithm.  
<a href="#_details">More...</a></p>
<code>#include &lt;stdio.h&gt;</code><br/>
<code>#include &lt;stdlib.h&gt;</code><br/>
<code>#include &lt;math.h&gt;</code><br/>
<code>#include &lt;ctype.h&gt;</code><br/>
<code>#include &lt;string.h&gt;</code><br/>
<code>#include &quot;io_lib_header.h&quot;</code><br/>
<code>#include &quot;util_lib_header.h&quot;</code><br/>
<code>#include &quot;define_header.h&quot;</code><br/>
<code>#include &quot;dp_lib_header.h&quot;</code><br/>
<code>#include &quot;fastal_lib_header.h&quot;</code><br/>
<code>#include &quot;fast_tree_header.h&quot;</code><br/>
<table border="0" cellpadding="0" cellspacing="0">
<tr><td colspan="2"><h2>Functions</h2></td></tr>
<tr><td class="memItemLeft" align="right" valign="top">int&nbsp;</td><td class="memItemRight" valign="bottom"><a class="el" href="fastal_8c.html#afbe437cab15b82544b8fc62221005018">sparse_dyn</a> (<a class="el" href="structFastal__profile.html">Fastal_profile</a> **profiles, <a class="el" href="structFastal__param.html">Fastal_param</a> *param_set, void *method_arguments_p, int is_dna, FILE *edit_file, FILE *prof_file, int number)</td></tr>
<tr><td class="mdescLeft">&nbsp;</td><td class="mdescRight">One run of sparse dynamic programming.  <a href="#afbe437cab15b82544b8fc62221005018"></a><br/></td></tr>
<tr><td class="memItemLeft" align="right" valign="top">int **&nbsp;</td><td class="memItemRight" valign="bottom"><a class="el" href="fastal_8c.html#aeac20b3b1749915e7d334e01d056417a">diagonals2int</a> (int *diagonals, int num_diagonals, char *seq1, char *seq2, int *num_points, <a class="el" href="structFastal__param.html">Fastal_param</a> *param_set)</td></tr>
<tr><td class="mdescLeft">&nbsp;</td><td class="mdescRight">Makes a sorted list out of diagonals.  <a href="#aeac20b3b1749915e7d334e01d056417a"></a><br/></td></tr>
<tr><td class="memItemLeft" align="right" valign="top">int **&nbsp;</td><td class="memItemRight" valign="bottom"><a class="el" href="fastal_8c.html#ab0a835bfe3d69a24b41c3dd353dda092">diagonals2int_gap_test</a> (int *diagonals, int num_diagonals, char *seq1, char *seq2, <a class="el" href="structFastal__profile.html">Fastal_profile</a> *profile1, <a class="el" href="structFastal__profile.html">Fastal_profile</a> *profile2, int *num_points, <a class="el" href="structFastal__param.html">Fastal_param</a> *param_set)</td></tr>
<tr><td class="mdescLeft">&nbsp;</td><td class="mdescRight">Makes a sorted list out of diagonals.  <a href="#ab0a835bfe3d69a24b41c3dd353dda092"></a><br/></td></tr>
<tr><td class="memItemLeft" align="right" valign="top">int&nbsp;</td><td class="memItemRight" valign="bottom"><a class="el" href="fastal_8c.html#ad5244a68c6325c1e3bb88195cf951b89">list2linked_pair_wise_fastal</a> (<a class="el" href="structFastal__profile.html">Fastal_profile</a> *prf1, <a class="el" href="structFastal__profile.html">Fastal_profile</a> *prf2, <a class="el" href="structFastal__param.html">Fastal_param</a> *param_set, int **list, int n, FILE *edit_f, FILE *prof_f, int node_number)</td></tr>
<tr><td class="mdescLeft">&nbsp;</td><td class="mdescRight">Calculates a fast and sparse dynamic programming matrix.  <a href="#ad5244a68c6325c1e3bb88195cf951b89"></a><br/></td></tr>
<tr><td class="memItemLeft" align="right" valign="top">char *&nbsp;</td><td class="memItemRight" valign="bottom"><a class="el" href="fastal_8c.html#a9bbdae2365a69e1783dbbf9b3ff3ba65">profile2consensus</a> (<a class="el" href="structFastal__profile.html">Fastal_profile</a> *profile, <a class="el" href="structFastal__param.html">Fastal_param</a> *param_set)</td></tr>
<tr><td class="mdescLeft">&nbsp;</td><td class="mdescRight">Turns a profile into a consensus sequence.  <a href="#a9bbdae2365a69e1783dbbf9b3ff3ba65"></a><br/></td></tr>
<tr><td class="memItemLeft" align="right" valign="top">int&nbsp;</td><td class="memItemRight" valign="bottom"><a class="el" href="fastal_8c.html#adb7e15a5c6a9c998e952ec110ba8d45b">seq_pair2diagonal_own</a> (char *seq1, char *seq2, int **diagonals, int *dig_length, int l1, int l2, int is_dna, int word_length)</td></tr>
<tr><td class="mdescLeft">&nbsp;</td><td class="mdescRight">Calculates the diagonals between two sequences.  <a href="#adb7e15a5c6a9c998e952ec110ba8d45b"></a><br/></td></tr>
<tr><td class="memItemLeft" align="right" valign="top">int&nbsp;</td><td class="memItemRight" valign="bottom"><a class="el" href="fastal_8c.html#ad21db2c5bf21fdebaf4367bcec2165fb">seq_pair2blast_diagonal</a> (char *seq_file_name1, char *seq_file_name2, int **diagonals, int *dig_length, int l1, int l2, int is_dna)</td></tr>
<tr><td class="mdescLeft">&nbsp;</td><td class="mdescRight">Calculates the diagonals between two sequences.  <a href="#ad21db2c5bf21fdebaf4367bcec2165fb"></a><br/></td></tr>
<tr><td class="memItemLeft" align="right" valign="top">int&nbsp;</td><td class="memItemRight" valign="bottom"><a class="el" href="fastal_8c.html#af19c3e4bd9a785b7904bd40249817880">seq_pair2blastz_diagonal</a> (char *seq_file_name1, char *seq_file_name2, int **diagonals, int *dig_length, int l1, int l2, int is_dna)</td></tr>
<tr><td class="mdescLeft">&nbsp;</td><td class="mdescRight">Calculates the diagonals between two sequences.  <a href="#af19c3e4bd9a785b7904bd40249817880"></a><br/></td></tr>
<tr><td class="memItemLeft" align="right" valign="top">int&nbsp;</td><td class="memItemRight" valign="bottom"><a class="el" href="fastal_8c.html#a9c2e3df348e18577cc3828b1dd94d70e">nw_dyn</a> (<a class="el" href="structFastal__profile.html">Fastal_profile</a> **profiles, <a class="el" href="structFastal__param.html">Fastal_param</a> *param_set, void *method_arguments_p, int is_dna, FILE *edit_file, FILE *prof_file, int number)</td></tr>
<tr><td class="mdescLeft">&nbsp;</td><td class="mdescRight">One run of needleman-wunsch dynamic programming.  <a href="#a9c2e3df348e18577cc3828b1dd94d70e"></a><br/></td></tr>
<tr><td class="memItemLeft" align="right" valign="top">int **&nbsp;</td><td class="memItemRight" valign="bottom"><a class="el" href="fastal_8c.html#a72b967437b83b6ea91c93ff204415add">sumup_profile</a> (<a class="el" href="structFastal__profile.html">Fastal_profile</a> *profile, int **sumup, <a class="el" href="structFastal__param.html">Fastal_param</a> *param_set)</td></tr>
<tr><td class="mdescLeft">&nbsp;</td><td class="mdescRight">This method takes a profile and turns it into a sumed up version.  <a href="#a72b967437b83b6ea91c93ff204415add"></a><br/></td></tr>
<tr><td class="memItemLeft" align="right" valign="top">int&nbsp;</td><td class="memItemRight" valign="bottom"><a class="el" href="fastal_8c.html#ac86132390d4a303d8619a5ba1f16f8fc">nw_matrix2edit_file</a> (double **prog_matrix, <a class="el" href="structFastal__profile.html">Fastal_profile</a> *prf1, <a class="el" href="structFastal__profile.html">Fastal_profile</a> *prf2, FILE *edit_f, int **prf_field, int *field_length, <a class="el" href="structFastal__param.html">Fastal_param</a> *param_set)</td></tr>
<tr><td class="mdescLeft">&nbsp;</td><td class="mdescRight">Turns the dynamic programming matrix into a editfile and calculates the new profile.  <a href="#ac86132390d4a303d8619a5ba1f16f8fc"></a><br/></td></tr>
<tr><td class="memItemLeft" align="right" valign="top">int&nbsp;</td><td class="memItemRight" valign="bottom"><a class="el" href="fastal_8c.html#ac021b7c48043748281d75b1f7f6eb230">prf_nw</a> (<a class="el" href="structFastal__profile.html">Fastal_profile</a> *profile1, <a class="el" href="structFastal__profile.html">Fastal_profile</a> *profile2, double **prog_matrix, FILE *edit_file_name, int **sumup_prf, int *sumup_length, <a class="el" href="structFastal__param.html">Fastal_param</a> *param_set)</td></tr>
<tr><td class="mdescLeft">&nbsp;</td><td class="mdescRight">Pairwise alignments of profile is done here.  <a href="#ac021b7c48043748281d75b1f7f6eb230"></a><br/></td></tr>
<tr><td class="memItemLeft" align="right" valign="top">int&nbsp;</td><td class="memItemRight" valign="bottom"><a class="el" href="fastal_8c.html#a56be015911a5d5349cc5b8a0fb230969">prf_gotoh</a> (<a class="el" href="structFastal__profile.html">Fastal_profile</a> *profile1, <a class="el" href="structFastal__profile.html">Fastal_profile</a> *profile2, FILE *edit_file_name, <a class="el" href="structGotoh__param.html">Gotoh_param</a> *arguments, <a class="el" href="structFastal__param.html">Fastal_param</a> *param_set)</td></tr>
<tr><td class="mdescLeft">&nbsp;</td><td class="mdescRight">The gotoh dynamic programming algorithm.  <a href="#a56be015911a5d5349cc5b8a0fb230969"></a><br/></td></tr>
<tr><td class="memItemLeft" align="right" valign="top">void&nbsp;</td><td class="memItemRight" valign="bottom"><a class="el" href="fastal_8c.html#a0605d5d64c57394977ca126191854d3e">alignment2files</a> (<a class="el" href="structFastal__profile.html">Fastal_profile</a> **profiles, <a class="el" href="structFastal__param.html">Fastal_param</a> *param_set, int **alignment, int alignment_length, FILE *edit_f, FILE *prof_f, int node_number)</td></tr>
<tr><td class="mdescLeft">&nbsp;</td><td class="mdescRight">Writes the alignment into the profile file and the edit file.  <a href="#a0605d5d64c57394977ca126191854d3e"></a><br/></td></tr>
<tr><td class="memItemLeft" align="right" valign="top">void&nbsp;</td><td class="memItemRight" valign="bottom"><a class="el" href="fastal_8c.html#a9bc5be0b8e5e0cbf9a8d142dfa6ba569">file_pos2profile</a> (FILE *seq_file, long off_set, <a class="el" href="structFastal__profile.html">Fastal_profile</a> *profile, int prf_number, <a class="el" href="structFastal__param.html">Fastal_param</a> *param_set)</td></tr>
<tr><td class="mdescLeft">&nbsp;</td><td class="mdescRight">Reads the sequence from a given position in a fasta file and turns it into a profile.  <a href="#a9bc5be0b8e5e0cbf9a8d142dfa6ba569"></a><br/></td></tr>
<tr><td class="memItemLeft" align="right" valign="top">int&nbsp;</td><td class="memItemRight" valign="bottom"><a class="el" href="fastal_8c.html#aa8a8d83816689f2b312fb9e4a427620a">make_index_of_file</a> (char *file_name, long **file_positions)</td></tr>
<tr><td class="mdescLeft">&nbsp;</td><td class="mdescRight">Constructs index of fasta_file.  <a href="#aa8a8d83816689f2b312fb9e4a427620a"></a><br/></td></tr>
<tr><td class="memItemLeft" align="right" valign="top">void&nbsp;</td><td class="memItemRight" valign="bottom"><a class="el" href="fastal_8c.html#a6700692fd6990ea839664439d51e3fc2">profile_file2profile</a> (<a class="el" href="structFastal__profile.html">Fastal_profile</a> *prof, FILE *profile_f, long position, <a class="el" href="structFastal__param.html">Fastal_param</a> *param_set)</td></tr>
<tr><td class="mdescLeft">&nbsp;</td><td class="mdescRight">Reads a profile from a profile file.  <a href="#a6700692fd6990ea839664439d51e3fc2"></a><br/></td></tr>
<tr><td class="memItemLeft" align="right" valign="top">void&nbsp;</td><td class="memItemRight" valign="bottom"><a class="el" href="fastal_8c.html#a749f8649a546a21fd013258de929e49e">profile2file</a> (<a class="el" href="structFastal__profile.html">Fastal_profile</a> *profile, FILE *file, <a class="el" href="structFastal__param.html">Fastal_param</a> *param_set)</td></tr>
<tr><td class="mdescLeft">&nbsp;</td><td class="mdescRight">Writes a profile into a file.  <a href="#a749f8649a546a21fd013258de929e49e"></a><br/></td></tr>
<tr><td class="memItemLeft" align="right" valign="top">void&nbsp;</td><td class="memItemRight" valign="bottom"><a class="el" href="fastal_8c.html#a52cb414188e66b5ad1dc7426ed9c45dd">edit_seq2aligned_seq</a> (char *aligned_sequence, FILE *sequence_file, long sequence_position, FILE *alignment_file)</td></tr>
<tr><td class="mdescLeft">&nbsp;</td><td class="mdescRight">Reads the profile out of an alignment (NOT IN USE).  <a href="#a52cb414188e66b5ad1dc7426ed9c45dd"></a><br/></td></tr>
<tr><td class="memItemLeft" align="right" valign="top">void&nbsp;</td><td class="memItemRight" valign="bottom"><a class="el" href="fastal_8c.html#a612ec92f470b9cbda9167c78f0152cc3">edit2alignment</a> (FILE *sequence_file, long *seq_positions, FILE *edit_file, long *edit_positions, int node_number, int number_of_sequences, char *aligned_sequence, int alignment_length, FILE *edit_seq_file, int offset, FILE *alignment_file)</td></tr>
<tr><td class="mdescLeft">&nbsp;</td><td class="mdescRight">Recursive function to turn the edit_file into the alignment.  <a href="#a612ec92f470b9cbda9167c78f0152cc3"></a><br/></td></tr>
<tr><td class="memItemLeft" align="right" valign="top">void&nbsp;</td><td class="memItemRight" valign="bottom"><a class="el" href="fastal_8c.html#a4f3eeb11dc9d50f60d94497ddff1421d">write2file</a> (int **sumup_prf, int length, FILE *file, int number, int num_sequences, <a class="el" href="structFastal__param.html">Fastal_param</a> *param_set)</td></tr>
<tr><td class="mdescLeft">&nbsp;</td><td class="mdescRight">Writes a profile to a file.  <a href="#a4f3eeb11dc9d50f60d94497ddff1421d"></a><br/></td></tr>
<tr><td class="memItemLeft" align="right" valign="top"><a class="anchor" id="aa9c66bc55318e62544273983419ee51f"></a><!-- doxytag: member="fastal.c::fastal_main" ref="aa9c66bc55318e62544273983419ee51f" args="(int argc, char **argv)" -->
int&nbsp;</td><td class="memItemRight" valign="bottom"><a class="el" href="fastal_8c.html#aa9c66bc55318e62544273983419ee51f">fastal_main</a> (int argc, char **argv)</td></tr>
<tr><td class="mdescLeft">&nbsp;</td><td class="mdescRight">main of the fastal algorithm <br/></td></tr>
<tr><td class="memItemLeft" align="right" valign="top">double **&nbsp;</td><td class="memItemRight" valign="bottom"><a class="el" href="fastal_8c.html#a80fce105127b3f6ad2d8fb039ffd4650">resize_dyn_matrix</a> (double **dyn_matrix, int old_length1, int old_length2, int length1, int length2)</td></tr>
<tr><td class="mdescLeft">&nbsp;</td><td class="mdescRight">Enlargement of the dynamic programming matrix in case it is to small.  <a href="#a80fce105127b3f6ad2d8fb039ffd4650"></a><br/></td></tr>
<tr><td class="memItemLeft" align="right" valign="top">void&nbsp;</td><td class="memItemRight" valign="bottom"><a class="el" href="fastal_8c.html#ab503a51a01e5a1b0c34729f6aa41e457">free_dyn_matrix</a> (int length1, double **dyn_matrix)</td></tr>
<tr><td class="mdescLeft">&nbsp;</td><td class="mdescRight">Frees the memory of a dynamic programming matrix.  <a href="#ab503a51a01e5a1b0c34729f6aa41e457"></a><br/></td></tr>
<tr><td class="memItemLeft" align="right" valign="top">void&nbsp;</td><td class="memItemRight" valign="bottom"><a class="el" href="fastal_8c.html#a924ff6917c4dee4fb2b7085d3fe52629">initiate_profiles</a> (<a class="el" href="structFastal__profile.html">Fastal_profile</a> **profiles, <a class="el" href="structFastal__param.html">Fastal_param</a> *param_set)</td></tr>
<tr><td class="mdescLeft">&nbsp;</td><td class="mdescRight">Initialises the profiles with basic values.  <a href="#a924ff6917c4dee4fb2b7085d3fe52629"></a><br/></td></tr>
<tr><td class="memItemLeft" align="right" valign="top">void&nbsp;</td><td class="memItemRight" valign="bottom"><a class="el" href="fastal_8c.html#a92748b5ec54f6c1b626d1a671bc4190e">free_fastal_profile</a> (<a class="el" href="structFastal__profile.html">Fastal_profile</a> *profile, int alphabet_size)</td></tr>
<tr><td class="mdescLeft">&nbsp;</td><td class="mdescRight">frees all memory occupied by the profile  <a href="#a92748b5ec54f6c1b626d1a671bc4190e"></a><br/></td></tr>
<tr><td class="memItemLeft" align="right" valign="top">void&nbsp;</td><td class="memItemRight" valign="bottom"><a class="el" href="fastal_8c.html#a968f2c97dee7c710f57561e72977549d">fill_parameters</a> (int is_dna, <a class="el" href="structFastal__param.html">Fastal_param</a> *param_set, char *method, char *diag_method, char *mat)</td></tr>
<tr><td class="mdescLeft">&nbsp;</td><td class="mdescRight">Initialize the Fastal parameter set.  <a href="#a968f2c97dee7c710f57561e72977549d"></a><br/></td></tr>
</table>
<hr/><a name="_details"></a><h2>Detailed Description</h2>
<p>Source code for the fastal algorithm. </p>
<dl class="author"><dt><b>Author:</b></dt><dd>Carsten Kemena </dd></dl>
<hr/><h2>Function Documentation</h2>
<a class="anchor" id="a0605d5d64c57394977ca126191854d3e"></a><!-- doxytag: member="fastal.c::alignment2files" ref="a0605d5d64c57394977ca126191854d3e" args="(Fastal_profile **profiles, Fastal_param *param_set, int **alignment, int alignment_length, FILE *edit_f, FILE *prof_f, int node_number)" -->
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">void alignment2files </td>
          <td>(</td>
          <td class="paramtype"><a class="el" href="structFastal__profile.html">Fastal_profile</a> **&nbsp;</td>
          <td class="paramname"> <em>profiles</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype"><a class="el" href="structFastal__param.html">Fastal_param</a> *&nbsp;</td>
          <td class="paramname"> <em>param_set</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">int **&nbsp;</td>
          <td class="paramname"> <em>alignment</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">int&nbsp;</td>
          <td class="paramname"> <em>alignment_length</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">FILE *&nbsp;</td>
          <td class="paramname"> <em>edit_f</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">FILE *&nbsp;</td>
          <td class="paramname"> <em>prof_f</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">int&nbsp;</td>
          <td class="paramname"> <em>node_number</em></td><td>&nbsp;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td><td></td>
        </tr>
      </table>
</div>
<div class="memdoc">

<p>Writes the alignment into the profile file and the edit file. </p>
<dl><dt><b>Parameters:</b></dt><dd>
  <table border="0" cellspacing="2" cellpadding="0">
    <tr><td valign="top"></td><td valign="top"><em>profiles</em>&nbsp;</td><td>The two profiles to combine. </td></tr>
    <tr><td valign="top"></td><td valign="top"><em>alignment</em>&nbsp;</td><td>The alinment information. </td></tr>
    <tr><td valign="top"></td><td valign="top"><em>alignment</em>&nbsp;</td><td>The length of the alignment. </td></tr>
    <tr><td valign="top"></td><td valign="top"><em>edit_f</em>&nbsp;</td><td>The edit file. </td></tr>
    <tr><td valign="top"></td><td valign="top"><em>prof_f</em>&nbsp;</td><td>The profile file. </td></tr>
    <tr><td valign="top"></td><td valign="top"><em>node_number</em>&nbsp;</td><td>the new node number. </td></tr>
  </table>
  </dd>
</dl>

</div>
</div>
<a class="anchor" id="aeac20b3b1749915e7d334e01d056417a"></a><!-- doxytag: member="fastal.c::diagonals2int" ref="aeac20b3b1749915e7d334e01d056417a" args="(int *diagonals, int num_diagonals, char *seq1, char *seq2, int *num_points, Fastal_param *param_set)" -->
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">int** diagonals2int </td>
          <td>(</td>
          <td class="paramtype">int *&nbsp;</td>
          <td class="paramname"> <em>diagonals</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">int&nbsp;</td>
          <td class="paramname"> <em>num_diagonals</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">char *&nbsp;</td>
          <td class="paramname"> <em>seq1</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">char *&nbsp;</td>
          <td class="paramname"> <em>seq2</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">int *&nbsp;</td>
          <td class="paramname"> <em>num_points</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype"><a class="el" href="structFastal__param.html">Fastal_param</a> *&nbsp;</td>
          <td class="paramname"> <em>param_set</em></td><td>&nbsp;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td><td></td>
        </tr>
      </table>
</div>
<div class="memdoc">

<p>Makes a sorted list out of diagonals. </p>
<dl><dt><b>Parameters:</b></dt><dd>
  <table border="0" cellspacing="2" cellpadding="0">
    <tr><td valign="top"></td><td valign="top"><em>diagonals</em>&nbsp;</td><td>A list of diagonals to use during dynamic programming. </td></tr>
    <tr><td valign="top"></td><td valign="top"><em>num_diagonals</em>&nbsp;</td><td>Number of diagonals. </td></tr>
    <tr><td valign="top"></td><td valign="top"><em>seq1</em>&nbsp;</td><td><a class="el" href="structSequence.html" title="Container for several sequences and their attributes.">Sequence</a> 1. </td></tr>
    <tr><td valign="top"></td><td valign="top"><em>seq2</em>&nbsp;</td><td><a class="el" href="structSequence.html" title="Container for several sequences and their attributes.">Sequence</a> 2. </td></tr>
    <tr><td valign="top"></td><td valign="top"><em>num_points</em>&nbsp;</td><td>Number of points in the list </td></tr>
    <tr><td valign="top"></td><td valign="top"><em>param_set</em>&nbsp;</td><td>Fastal parameters. </td></tr>
  </table>
  </dd>
</dl>
<dl class="return"><dt><b>Returns:</b></dt><dd>A 2-dim array which contains all points needed for the sparse dynamic programming algorithm. </dd></dl>

</div>
</div>
<a class="anchor" id="ab0a835bfe3d69a24b41c3dd353dda092"></a><!-- doxytag: member="fastal.c::diagonals2int_gap_test" ref="ab0a835bfe3d69a24b41c3dd353dda092" args="(int *diagonals, int num_diagonals, char *seq1, char *seq2, Fastal_profile *profile1, Fastal_profile *profile2, int *num_points, Fastal_param *param_set)" -->
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">int** diagonals2int_gap_test </td>
          <td>(</td>
          <td class="paramtype">int *&nbsp;</td>
          <td class="paramname"> <em>diagonals</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">int&nbsp;</td>
          <td class="paramname"> <em>num_diagonals</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">char *&nbsp;</td>
          <td class="paramname"> <em>seq1</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">char *&nbsp;</td>
          <td class="paramname"> <em>seq2</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype"><a class="el" href="structFastal__profile.html">Fastal_profile</a> *&nbsp;</td>
          <td class="paramname"> <em>profile1</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype"><a class="el" href="structFastal__profile.html">Fastal_profile</a> *&nbsp;</td>
          <td class="paramname"> <em>profile2</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">int *&nbsp;</td>
          <td class="paramname"> <em>num_points</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype"><a class="el" href="structFastal__param.html">Fastal_param</a> *&nbsp;</td>
          <td class="paramname"> <em>param_set</em></td><td>&nbsp;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td><td></td>
        </tr>
      </table>
</div>
<div class="memdoc">

<p>Makes a sorted list out of diagonals. </p>
<dl><dt><b>Parameters:</b></dt><dd>
  <table border="0" cellspacing="2" cellpadding="0">
    <tr><td valign="top"></td><td valign="top"><em>diagonals</em>&nbsp;</td><td>A list of diagonals to use during dynamic programming. </td></tr>
    <tr><td valign="top"></td><td valign="top"><em>num_diagonals</em>&nbsp;</td><td>Number of diagonals. </td></tr>
    <tr><td valign="top"></td><td valign="top"><em>seq1</em>&nbsp;</td><td><a class="el" href="structSequence.html" title="Container for several sequences and their attributes.">Sequence</a> 1. </td></tr>
    <tr><td valign="top"></td><td valign="top"><em>seq2</em>&nbsp;</td><td><a class="el" href="structSequence.html" title="Container for several sequences and their attributes.">Sequence</a> 2. </td></tr>
    <tr><td valign="top"></td><td valign="top"><em>num_points</em>&nbsp;</td><td>Number of points in the list </td></tr>
    <tr><td valign="top"></td><td valign="top"><em>param_set</em>&nbsp;</td><td>Fastal parameters. </td></tr>
  </table>
  </dd>
</dl>
<dl class="return"><dt><b>Returns:</b></dt><dd>A 2-dim array which contains all points needed for the sparse dynamic programming algorithm. </dd></dl>

</div>
</div>
<a class="anchor" id="a612ec92f470b9cbda9167c78f0152cc3"></a><!-- doxytag: member="fastal.c::edit2alignment" ref="a612ec92f470b9cbda9167c78f0152cc3" args="(FILE *sequence_file, long *seq_positions, FILE *edit_file, long *edit_positions, int node_number, int number_of_sequences, char *aligned_sequence, int alignment_length, FILE *edit_seq_file, int offset, FILE *alignment_file)" -->
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">void edit2alignment </td>
          <td>(</td>
          <td class="paramtype">FILE *&nbsp;</td>
          <td class="paramname"> <em>sequence_file</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">long *&nbsp;</td>
          <td class="paramname"> <em>seq_positions</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">FILE *&nbsp;</td>
          <td class="paramname"> <em>edit_file</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">long *&nbsp;</td>
          <td class="paramname"> <em>edit_positions</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">int&nbsp;</td>
          <td class="paramname"> <em>node_number</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">int&nbsp;</td>
          <td class="paramname"> <em>number_of_sequences</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">char *&nbsp;</td>
          <td class="paramname"> <em>aligned_sequence</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">int&nbsp;</td>
          <td class="paramname"> <em>alignment_length</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">FILE *&nbsp;</td>
          <td class="paramname"> <em>edit_seq_file</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">int&nbsp;</td>
          <td class="paramname"> <em>offset</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">FILE *&nbsp;</td>
          <td class="paramname"> <em>alignment_file</em></td><td>&nbsp;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td><td></td>
        </tr>
      </table>
</div>
<div class="memdoc">

<p>Recursive function to turn the edit_file into the alignment. </p>
<dl><dt><b>Parameters:</b></dt><dd>
  <table border="0" cellspacing="2" cellpadding="0">
    <tr><td valign="top"></td><td valign="top"><em>sequence_file</em>&nbsp;</td><td>File with all sequences. </td></tr>
    <tr><td valign="top"></td><td valign="top"><em>sequence_position</em>&nbsp;</td><td>The array of sequence positions in <em>sequence_file</em> </td></tr>
    <tr><td valign="top"></td><td valign="top"><em>edit_file</em>&nbsp;</td><td>File to safe the edit profiles in. </td></tr>
    <tr><td valign="top"></td><td valign="top"><em>edit_positions</em>&nbsp;</td><td>Array saving the coorespondence between edit profile and position in <em>edit_file</em> </td></tr>
    <tr><td valign="top"></td><td valign="top"><em>node_number</em>&nbsp;</td><td>The current node. </td></tr>
    <tr><td valign="top"></td><td valign="top"><em>number_of_sequences</em>&nbsp;</td><td>The number of sequences. </td></tr>
    <tr><td valign="top"></td><td valign="top"><em>aligned_sequence</em>&nbsp;</td><td>The sequence that is edited. </td></tr>
    <tr><td valign="top"></td><td valign="top"><em>alignment_length</em>&nbsp;</td><td>The length of the alignment. </td></tr>
    <tr><td valign="top"></td><td valign="top"><em>edit_seq_file</em>&nbsp;</td><td>File that saves the edited_sequences of the internal nodes. </td></tr>
    <tr><td valign="top"></td><td valign="top"><em>offset</em>&nbsp;</td><td>Saves the size of the edited_sequences. </td></tr>
    <tr><td valign="top"></td><td valign="top"><em>alignment_file</em>&nbsp;</td><td>File where the alignment is saved. </td></tr>
  </table>
  </dd>
</dl>

</div>
</div>
<a class="anchor" id="a52cb414188e66b5ad1dc7426ed9c45dd"></a><!-- doxytag: member="fastal.c::edit_seq2aligned_seq" ref="a52cb414188e66b5ad1dc7426ed9c45dd" args="(char *aligned_sequence, FILE *sequence_file, long sequence_position, FILE *alignment_file)" -->
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">void edit_seq2aligned_seq </td>
          <td>(</td>
          <td class="paramtype">char *&nbsp;</td>
          <td class="paramname"> <em>aligned_sequence</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">FILE *&nbsp;</td>
          <td class="paramname"> <em>sequence_file</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">long&nbsp;</td>
          <td class="paramname"> <em>sequence_position</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">FILE *&nbsp;</td>
          <td class="paramname"> <em>alignment_file</em></td><td>&nbsp;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td><td></td>
        </tr>
      </table>
</div>
<div class="memdoc">

<p>Reads the profile out of an alignment (NOT IN USE). </p>
<p>Writes the sequence into the alignment_file.</p>
<dl><dt><b>Parameters:</b></dt><dd>
  <table border="0" cellspacing="2" cellpadding="0">
    <tr><td valign="top"></td><td valign="top"><em>aligned_sequence</em>&nbsp;</td><td>Pattern of aligned sequence. </td></tr>
    <tr><td valign="top"></td><td valign="top"><em>sequence_file</em>&nbsp;</td><td>File with sequences. </td></tr>
    <tr><td valign="top"></td><td valign="top"><em>sequence_position</em>&nbsp;</td><td>Positions of sequences in <em>sequence_file</em>. </td></tr>
    <tr><td valign="top"></td><td valign="top"><em>alignment_file</em>&nbsp;</td><td>The file to write the sequence into. </td></tr>
  </table>
  </dd>
</dl>

</div>
</div>
<a class="anchor" id="a9bc5be0b8e5e0cbf9a8d142dfa6ba569"></a><!-- doxytag: member="fastal.c::file_pos2profile" ref="a9bc5be0b8e5e0cbf9a8d142dfa6ba569" args="(FILE *seq_file, long off_set, Fastal_profile *profile, int prf_number, Fastal_param *param_set)" -->
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">void file_pos2profile </td>
          <td>(</td>
          <td class="paramtype">FILE *&nbsp;</td>
          <td class="paramname"> <em>seq_file</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">long&nbsp;</td>
          <td class="paramname"> <em>off_set</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype"><a class="el" href="structFastal__profile.html">Fastal_profile</a> *&nbsp;</td>
          <td class="paramname"> <em>profile</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">int&nbsp;</td>
          <td class="paramname"> <em>prf_number</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype"><a class="el" href="structFastal__param.html">Fastal_param</a> *&nbsp;</td>
          <td class="paramname"> <em>param_set</em></td><td>&nbsp;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td><td></td>
        </tr>
      </table>
</div>
<div class="memdoc">

<p>Reads the sequence from a given position in a fasta file and turns it into a profile. </p>
<dl><dt><b>Parameters:</b></dt><dd>
  <table border="0" cellspacing="2" cellpadding="0">
    <tr><td valign="top"></td><td valign="top"><em>seq_file</em>&nbsp;</td><td>The file where the sequence is stored. </td></tr>
    <tr><td valign="top"></td><td valign="top"><em>off_set</em>&nbsp;</td><td>The off_set from the beginning of the file to the position of the sequence name. </td></tr>
    <tr><td valign="top"></td><td valign="top"><em>profile</em>&nbsp;</td><td>The profile where the sequence will be stored into. </td></tr>
    <tr><td valign="top"></td><td valign="top"><em>prf_number</em>&nbsp;</td><td>The number of this profile. </td></tr>
  </table>
  </dd>
</dl>

</div>
</div>
<a class="anchor" id="a968f2c97dee7c710f57561e72977549d"></a><!-- doxytag: member="fastal.c::fill_parameters" ref="a968f2c97dee7c710f57561e72977549d" args="(int is_dna, Fastal_param *param_set, char *method, char *diag_method, char *mat)" -->
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">void fill_parameters </td>
          <td>(</td>
          <td class="paramtype">int&nbsp;</td>
          <td class="paramname"> <em>is_dna</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype"><a class="el" href="structFastal__param.html">Fastal_param</a> *&nbsp;</td>
          <td class="paramname"> <em>param_set</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">char *&nbsp;</td>
          <td class="paramname"> <em>method</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">char *&nbsp;</td>
          <td class="paramname"> <em>diag_method</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">char *&nbsp;</td>
          <td class="paramname"> <em>mat</em></td><td>&nbsp;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td><td></td>
        </tr>
      </table>
</div>
<div class="memdoc">

<p>Initialize the Fastal parameter set. </p>
<dl><dt><b>Parameters:</b></dt><dd>
  <table border="0" cellspacing="2" cellpadding="0">
    <tr><td valign="top"></td><td valign="top"><em>is_dna</em>&nbsp;</td><td>1 when sequences are dna, 0 when not </td></tr>
    <tr><td valign="top"></td><td valign="top"><em>param_set</em>&nbsp;</td><td>The fastal parameter set. </td></tr>
    <tr><td valign="top"></td><td valign="top"><em>method</em>&nbsp;</td><td>The method to use in Fastal. </td></tr>
  </table>
  </dd>
</dl>

</div>
</div>
<a class="anchor" id="ab503a51a01e5a1b0c34729f6aa41e457"></a><!-- doxytag: member="fastal.c::free_dyn_matrix" ref="ab503a51a01e5a1b0c34729f6aa41e457" args="(int length1, double **dyn_matrix)" -->
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">void free_dyn_matrix </td>
          <td>(</td>
          <td class="paramtype">int&nbsp;</td>
          <td class="paramname"> <em>length1</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">double **&nbsp;</td>
          <td class="paramname"> <em>dyn_matrix</em></td><td>&nbsp;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td><td></td>
        </tr>
      </table>
</div>
<div class="memdoc">

<p>Frees the memory of a dynamic programming matrix. </p>
<dl><dt><b>Parameters:</b></dt><dd>
  <table border="0" cellspacing="2" cellpadding="0">
    <tr><td valign="top"></td><td valign="top"><em>length1</em>&nbsp;</td><td>Size of the first dimension of the matrix. </td></tr>
    <tr><td valign="top"></td><td valign="top"><em>dyn_matrix</em>&nbsp;</td><td>The dynamic programming matrix. </td></tr>
  </table>
  </dd>
</dl>

</div>
</div>
<a class="anchor" id="a92748b5ec54f6c1b626d1a671bc4190e"></a><!-- doxytag: member="fastal.c::free_fastal_profile" ref="a92748b5ec54f6c1b626d1a671bc4190e" args="(Fastal_profile *profile, int alphabet_size)" -->
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">void free_fastal_profile </td>
          <td>(</td>
          <td class="paramtype"><a class="el" href="structFastal__profile.html">Fastal_profile</a> *&nbsp;</td>
          <td class="paramname"> <em>profile</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">int&nbsp;</td>
          <td class="paramname"> <em>alphabet_size</em></td><td>&nbsp;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td><td></td>
        </tr>
      </table>
</div>
<div class="memdoc">

<p>frees all memory occupied by the profile </p>
<dl><dt><b>Parameters:</b></dt><dd>
  <table border="0" cellspacing="2" cellpadding="0">
    <tr><td valign="top"></td><td valign="top"><em>profile</em>&nbsp;</td><td>The profile to free. </td></tr>
    <tr><td valign="top"></td><td valign="top"><em>alphabet_size</em>&nbsp;</td><td>The alphabet_size. </td></tr>
  </table>
  </dd>
</dl>

</div>
</div>
<a class="anchor" id="a924ff6917c4dee4fb2b7085d3fe52629"></a><!-- doxytag: member="fastal.c::initiate_profiles" ref="a924ff6917c4dee4fb2b7085d3fe52629" args="(Fastal_profile **profiles, Fastal_param *param_set)" -->
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">void initiate_profiles </td>
          <td>(</td>
          <td class="paramtype"><a class="el" href="structFastal__profile.html">Fastal_profile</a> **&nbsp;</td>
          <td class="paramname"> <em>profiles</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype"><a class="el" href="structFastal__param.html">Fastal_param</a> *&nbsp;</td>
          <td class="paramname"> <em>param_set</em></td><td>&nbsp;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td><td></td>
        </tr>
      </table>
</div>
<div class="memdoc">

<p>Initialises the profiles with basic values. </p>
<dl><dt><b>Parameters:</b></dt><dd>
  <table border="0" cellspacing="2" cellpadding="0">
    <tr><td valign="top"></td><td valign="top"><em>profiles</em>&nbsp;</td><td>Array of 3 profiles. </td></tr>
    <tr><td valign="top"></td><td valign="top"><em>param_set</em>&nbsp;</td><td>The fastal parameters </td></tr>
  </table>
  </dd>
</dl>

</div>
</div>
<a class="anchor" id="ad5244a68c6325c1e3bb88195cf951b89"></a><!-- doxytag: member="fastal.c::list2linked_pair_wise_fastal" ref="ad5244a68c6325c1e3bb88195cf951b89" args="(Fastal_profile *prf1, Fastal_profile *prf2, Fastal_param *param_set, int **list, int n, FILE *edit_f, FILE *prof_f, int node_number)" -->
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">int list2linked_pair_wise_fastal </td>
          <td>(</td>
          <td class="paramtype"><a class="el" href="structFastal__profile.html">Fastal_profile</a> *&nbsp;</td>
          <td class="paramname"> <em>prf1</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype"><a class="el" href="structFastal__profile.html">Fastal_profile</a> *&nbsp;</td>
          <td class="paramname"> <em>prf2</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype"><a class="el" href="structFastal__param.html">Fastal_param</a> *&nbsp;</td>
          <td class="paramname"> <em>param_set</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">int **&nbsp;</td>
          <td class="paramname"> <em>list</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">int&nbsp;</td>
          <td class="paramname"> <em>n</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">FILE *&nbsp;</td>
          <td class="paramname"> <em>edit_f</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">FILE *&nbsp;</td>
          <td class="paramname"> <em>prof_f</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">int&nbsp;</td>
          <td class="paramname"> <em>node_number</em></td><td>&nbsp;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td><td></td>
        </tr>
      </table>
</div>
<div class="memdoc">

<p>Calculates a fast and sparse dynamic programming matrix. </p>
<dl><dt><b>Parameters:</b></dt><dd>
  <table border="0" cellspacing="2" cellpadding="0">
    <tr><td valign="top"></td><td valign="top"><em>prf1</em>&nbsp;</td><td>Profile of first sequence. </td></tr>
    <tr><td valign="top"></td><td valign="top"><em>prf2</em>&nbsp;</td><td>Profile of second sequence. </td></tr>
    <tr><td valign="top"></td><td valign="top"><em>param_set</em>&nbsp;</td><td>The parameter for the alignment. </td></tr>
    <tr><td valign="top"></td><td valign="top"><em>list</em>&nbsp;</td><td>The list of diagonals. </td></tr>
    <tr><td valign="top"></td><td valign="top"><em>n</em>&nbsp;</td><td>number of dots. </td></tr>
    <tr><td valign="top"></td><td valign="top"><em>edit_f</em>&nbsp;</td><td>File to save the edit information. </td></tr>
    <tr><td valign="top"></td><td valign="top"><em>prof_f</em>&nbsp;</td><td>File to save the profile. </td></tr>
    <tr><td valign="top"></td><td valign="top"><em>node_number</em>&nbsp;</td><td>Number of the new profile. </td></tr>
  </table>
  </dd>
</dl>

</div>
</div>
<a class="anchor" id="aa8a8d83816689f2b312fb9e4a427620a"></a><!-- doxytag: member="fastal.c::make_index_of_file" ref="aa8a8d83816689f2b312fb9e4a427620a" args="(char *file_name, long **file_positions)" -->
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">int make_index_of_file </td>
          <td>(</td>
          <td class="paramtype">char *&nbsp;</td>
          <td class="paramname"> <em>file_name</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">long **&nbsp;</td>
          <td class="paramname"> <em>file_positions</em></td><td>&nbsp;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td><td></td>
        </tr>
      </table>
</div>
<div class="memdoc">

<p>Constructs index of fasta_file. </p>
<p>The index is of length n (n= number of sequences in the given multi fasta file.). In the order of appearance in the file the position of each sequence in the file is stored. </p>
<dl><dt><b>Parameters:</b></dt><dd>
  <table border="0" cellspacing="2" cellpadding="0">
    <tr><td valign="top"></td><td valign="top"><em>file_name</em>&nbsp;</td><td>The file with the sequences. </td></tr>
    <tr><td valign="top"></td><td valign="top"><em>file_positions</em>&nbsp;</td><td>Array to save the positions in. </td></tr>
  </table>
  </dd>
</dl>
<dl class="return"><dt><b>Returns:</b></dt><dd>The number of sequences in <em>file_name</em>. </dd></dl>

</div>
</div>
<a class="anchor" id="a9c2e3df348e18577cc3828b1dd94d70e"></a><!-- doxytag: member="fastal.c::nw_dyn" ref="a9c2e3df348e18577cc3828b1dd94d70e" args="(Fastal_profile **profiles, Fastal_param *param_set, void *method_arguments_p, int is_dna, FILE *edit_file, FILE *prof_file, int number)" -->
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">int nw_dyn </td>
          <td>(</td>
          <td class="paramtype"><a class="el" href="structFastal__profile.html">Fastal_profile</a> **&nbsp;</td>
          <td class="paramname"> <em>profiles</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype"><a class="el" href="structFastal__param.html">Fastal_param</a> *&nbsp;</td>
          <td class="paramname"> <em>param_set</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">void *&nbsp;</td>
          <td class="paramname"> <em>method_arguments_p</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">int&nbsp;</td>
          <td class="paramname"> <em>is_dna</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">FILE *&nbsp;</td>
          <td class="paramname"> <em>edit_file</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">FILE *&nbsp;</td>
          <td class="paramname"> <em>prof_file</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">int&nbsp;</td>
          <td class="paramname"> <em>number</em></td><td>&nbsp;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td><td></td>
        </tr>
      </table>
</div>
<div class="memdoc">

<p>One run of needleman-wunsch dynamic programming. </p>
<dl><dt><b>Parameters:</b></dt><dd>
  <table border="0" cellspacing="2" cellpadding="0">
    <tr><td valign="top"></td><td valign="top"><em>profiles</em>&nbsp;</td><td>The profiles. </td></tr>
    <tr><td valign="top"></td><td valign="top"><em>param_set</em>&nbsp;</td><td>The fastal parameters. </td></tr>
    <tr><td valign="top"></td><td valign="top"><em>method_arguments_p</em>&nbsp;</td><td>The method arguments. </td></tr>
    <tr><td valign="top"></td><td valign="top"><em>is_dna</em>&nbsp;</td><td>Sequences are DNA (<em>is_dna</em> = 1) or protein. </td></tr>
    <tr><td valign="top"></td><td valign="top"><em>edit_file</em>&nbsp;</td><td>The edit file. </td></tr>
    <tr><td valign="top"></td><td valign="top"><em>prof_file</em>&nbsp;</td><td>the profile file. </td></tr>
    <tr><td valign="top"></td><td valign="top"><em>number</em>&nbsp;</td><td>Number of the parent node. </td></tr>
  </table>
  </dd>
</dl>
<dl class="return"><dt><b>Returns:</b></dt><dd>The length of the alignment. </dd></dl>

</div>
</div>
<a class="anchor" id="ac86132390d4a303d8619a5ba1f16f8fc"></a><!-- doxytag: member="fastal.c::nw_matrix2edit_file" ref="ac86132390d4a303d8619a5ba1f16f8fc" args="(double **prog_matrix, Fastal_profile *prf1, Fastal_profile *prf2, FILE *edit_f, int **prf_field, int *field_length, Fastal_param *param_set)" -->
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">int nw_matrix2edit_file </td>
          <td>(</td>
          <td class="paramtype">double **&nbsp;</td>
          <td class="paramname"> <em>prog_matrix</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype"><a class="el" href="structFastal__profile.html">Fastal_profile</a> *&nbsp;</td>
          <td class="paramname"> <em>prf1</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype"><a class="el" href="structFastal__profile.html">Fastal_profile</a> *&nbsp;</td>
          <td class="paramname"> <em>prf2</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">FILE *&nbsp;</td>
          <td class="paramname"> <em>edit_f</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">int **&nbsp;</td>
          <td class="paramname"> <em>prf_field</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">int *&nbsp;</td>
          <td class="paramname"> <em>field_length</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype"><a class="el" href="structFastal__param.html">Fastal_param</a> *&nbsp;</td>
          <td class="paramname"> <em>param_set</em></td><td>&nbsp;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td><td></td>
        </tr>
      </table>
</div>
<div class="memdoc">

<p>Turns the dynamic programming matrix into a editfile and calculates the new profile. </p>
<p>Required for NW-algorithm. </p>
<dl><dt><b>Parameters:</b></dt><dd>
  <table border="0" cellspacing="2" cellpadding="0">
    <tr><td valign="top"></td><td valign="top"><em>prog_matrix</em>&nbsp;</td><td>The dynamic programming matrix. </td></tr>
    <tr><td valign="top"></td><td valign="top"><em>prf1</em>&nbsp;</td><td>The first profile. </td></tr>
    <tr><td valign="top"></td><td valign="top"><em>prf2</em>&nbsp;</td><td>The second profile. </td></tr>
    <tr><td valign="top"></td><td valign="top"><em>edit_f</em>&nbsp;</td><td>A File object (already opened) to write the edit sequence into. </td></tr>
    <tr><td valign="top"></td><td valign="top"><em>prf_field</em>&nbsp;</td><td>A 2-dim array to save the new profile into. </td></tr>
    <tr><td valign="top"></td><td valign="top"><em>field_length</em>&nbsp;</td><td>Length of the new profile. </td></tr>
    <tr><td valign="top"></td><td valign="top"><em>param_set</em>&nbsp;</td><td>Parameters of the Fastal-Algorithm </td></tr>
  </table>
  </dd>
</dl>

</div>
</div>
<a class="anchor" id="a56be015911a5d5349cc5b8a0fb230969"></a><!-- doxytag: member="fastal.c::prf_gotoh" ref="a56be015911a5d5349cc5b8a0fb230969" args="(Fastal_profile *profile1, Fastal_profile *profile2, FILE *edit_file_name, Gotoh_param *arguments, Fastal_param *param_set)" -->
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">int prf_gotoh </td>
          <td>(</td>
          <td class="paramtype"><a class="el" href="structFastal__profile.html">Fastal_profile</a> *&nbsp;</td>
          <td class="paramname"> <em>profile1</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype"><a class="el" href="structFastal__profile.html">Fastal_profile</a> *&nbsp;</td>
          <td class="paramname"> <em>profile2</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">FILE *&nbsp;</td>
          <td class="paramname"> <em>edit_file_name</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype"><a class="el" href="structGotoh__param.html">Gotoh_param</a> *&nbsp;</td>
          <td class="paramname"> <em>arguments</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype"><a class="el" href="structFastal__param.html">Fastal_param</a> *&nbsp;</td>
          <td class="paramname"> <em>param_set</em></td><td>&nbsp;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td><td></td>
        </tr>
      </table>
</div>
<div class="memdoc">

<p>The gotoh dynamic programming algorithm. </p>
<dl><dt><b>Parameters:</b></dt><dd>
  <table border="0" cellspacing="2" cellpadding="0">
    <tr><td valign="top"></td><td valign="top"><em>profile1</em>&nbsp;</td><td>The first profile. </td></tr>
  </table>
  </dd>
</dl>

</div>
</div>
<a class="anchor" id="ac021b7c48043748281d75b1f7f6eb230"></a><!-- doxytag: member="fastal.c::prf_nw" ref="ac021b7c48043748281d75b1f7f6eb230" args="(Fastal_profile *profile1, Fastal_profile *profile2, double **prog_matrix, FILE *edit_file_name, int **sumup_prf, int *sumup_length, Fastal_param *param_set)" -->
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">int prf_nw </td>
          <td>(</td>
          <td class="paramtype"><a class="el" href="structFastal__profile.html">Fastal_profile</a> *&nbsp;</td>
          <td class="paramname"> <em>profile1</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype"><a class="el" href="structFastal__profile.html">Fastal_profile</a> *&nbsp;</td>
          <td class="paramname"> <em>profile2</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">double **&nbsp;</td>
          <td class="paramname"> <em>prog_matrix</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">FILE *&nbsp;</td>
          <td class="paramname"> <em>edit_file_name</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">int **&nbsp;</td>
          <td class="paramname"> <em>sumup_prf</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">int *&nbsp;</td>
          <td class="paramname"> <em>sumup_length</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype"><a class="el" href="structFastal__param.html">Fastal_param</a> *&nbsp;</td>
          <td class="paramname"> <em>param_set</em></td><td>&nbsp;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td><td></td>
        </tr>
      </table>
</div>
<div class="memdoc">

<p>Pairwise alignments of profile is done here. </p>
<dl><dt><b>Parameters:</b></dt><dd>
  <table border="0" cellspacing="2" cellpadding="0">
    <tr><td valign="top"></td><td valign="top"><em>profile1</em>&nbsp;</td><td>Profile of sequence 1 </td></tr>
    <tr><td valign="top"></td><td valign="top"><em>profile2</em>&nbsp;</td><td>Profile of sequence 2 </td></tr>
    <tr><td valign="top"></td><td valign="top"><em>prog_matrix</em>&nbsp;</td><td>Matrix for dynamic programming </td></tr>
    <tr><td valign="top"></td><td valign="top"><em>edit_file_name</em>&nbsp;</td><td>The edit_file_name </td></tr>
    <tr><td valign="top"></td><td valign="top"><em>sumup_prf</em>&nbsp;</td><td>The sumup version of profile 1, which later contains the aligned profile. </td></tr>
    <tr><td valign="top"></td><td valign="top"><em>sumup_length</em>&nbsp;</td><td>Contains the length of the aligned profile. </td></tr>
  </table>
  </dd>
</dl>
<dl class="return"><dt><b>Returns:</b></dt><dd>length of the aligned profile </dd></dl>

</div>
</div>
<a class="anchor" id="a9bbdae2365a69e1783dbbf9b3ff3ba65"></a><!-- doxytag: member="fastal.c::profile2consensus" ref="a9bbdae2365a69e1783dbbf9b3ff3ba65" args="(Fastal_profile *profile, Fastal_param *param_set)" -->
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">char* profile2consensus </td>
          <td>(</td>
          <td class="paramtype"><a class="el" href="structFastal__profile.html">Fastal_profile</a> *&nbsp;</td>
          <td class="paramname"> <em>profile</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype"><a class="el" href="structFastal__param.html">Fastal_param</a> *&nbsp;</td>
          <td class="paramname"> <em>param_set</em></td><td>&nbsp;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td><td></td>
        </tr>
      </table>
</div>
<div class="memdoc">

<p>Turns a profile into a consensus sequence. </p>
<p>The character with the highest number of occurences is used as consensus. Gaps are not included. For example: 10 '-' and one 'A' would give 'A' as consensus. </p>
<dl><dt><b>Parameters:</b></dt><dd>
  <table border="0" cellspacing="2" cellpadding="0">
    <tr><td valign="top"></td><td valign="top"><em>profile</em>&nbsp;</td><td>The profile. </td></tr>
    <tr><td valign="top"></td><td valign="top"><em>file_name</em>&nbsp;</td><td>Name of the file to save the consensus sequence in. </td></tr>
    <tr><td valign="top"></td><td valign="top"><em>param_set</em>&nbsp;</td><td>The parameter of the fastal algorithm. </td></tr>
  </table>
  </dd>
</dl>
<dl class="return"><dt><b>Returns:</b></dt><dd>the sequence </dd></dl>

</div>
</div>
<a class="anchor" id="a749f8649a546a21fd013258de929e49e"></a><!-- doxytag: member="fastal.c::profile2file" ref="a749f8649a546a21fd013258de929e49e" args="(Fastal_profile *profile, FILE *file, Fastal_param *param_set)" -->
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">void profile2file </td>
          <td>(</td>
          <td class="paramtype"><a class="el" href="structFastal__profile.html">Fastal_profile</a> *&nbsp;</td>
          <td class="paramname"> <em>profile</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">FILE *&nbsp;</td>
          <td class="paramname"> <em>file</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype"><a class="el" href="structFastal__param.html">Fastal_param</a> *&nbsp;</td>
          <td class="paramname"> <em>param_set</em></td><td>&nbsp;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td><td></td>
        </tr>
      </table>
</div>
<div class="memdoc">

<p>Writes a profile into a file. </p>
<dl><dt><b>Parameters:</b></dt><dd>
  <table border="0" cellspacing="2" cellpadding="0">
    <tr><td valign="top"></td><td valign="top"><em>profile</em>&nbsp;</td><td>Pointer to the profile which has to be saved. </td></tr>
    <tr><td valign="top"></td><td valign="top"><em>file</em>&nbsp;</td><td>A File object (already opened) to write the profile to. </td></tr>
    <tr><td valign="top"></td><td valign="top"><em>param_set</em>&nbsp;</td><td>The parameters for the fastal algorithm. </td></tr>
  </table>
  </dd>
</dl>

</div>
</div>
<a class="anchor" id="a6700692fd6990ea839664439d51e3fc2"></a><!-- doxytag: member="fastal.c::profile_file2profile" ref="a6700692fd6990ea839664439d51e3fc2" args="(Fastal_profile *prof, FILE *profile_f, long position, Fastal_param *param_set)" -->
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">void profile_file2profile </td>
          <td>(</td>
          <td class="paramtype"><a class="el" href="structFastal__profile.html">Fastal_profile</a> *&nbsp;</td>
          <td class="paramname"> <em>prof</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">FILE *&nbsp;</td>
          <td class="paramname"> <em>profile_f</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">long&nbsp;</td>
          <td class="paramname"> <em>position</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype"><a class="el" href="structFastal__param.html">Fastal_param</a> *&nbsp;</td>
          <td class="paramname"> <em>param_set</em></td><td>&nbsp;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td><td></td>
        </tr>
      </table>
</div>
<div class="memdoc">

<p>Reads a profile from a profile file. </p>
<dl><dt><b>Parameters:</b></dt><dd>
  <table border="0" cellspacing="2" cellpadding="0">
    <tr><td valign="top"></td><td valign="top"><em>prof</em>&nbsp;</td><td>A <a class="el" href="structFastal__profile.html" title="Struct to save all informations of a profile.">Fastal_profile</a> object to save the profile in. </td></tr>
    <tr><td valign="top"></td><td valign="top"><em>profile_f</em>&nbsp;</td><td>file where the profile is stored. </td></tr>
    <tr><td valign="top"></td><td valign="top"><em>position</em>&nbsp;</td><td>Position of the profile in <em>profile_f</em>. </td></tr>
    <tr><td valign="top"></td><td valign="top"><em>param_set</em>&nbsp;</td><td>The parameter set for Fastal </td></tr>
  </table>
  </dd>
</dl>

</div>
</div>
<a class="anchor" id="a80fce105127b3f6ad2d8fb039ffd4650"></a><!-- doxytag: member="fastal.c::resize_dyn_matrix" ref="a80fce105127b3f6ad2d8fb039ffd4650" args="(double **dyn_matrix, int old_length1, int old_length2, int length1, int length2)" -->
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">double** resize_dyn_matrix </td>
          <td>(</td>
          <td class="paramtype">double **&nbsp;</td>
          <td class="paramname"> <em>dyn_matrix</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">int&nbsp;</td>
          <td class="paramname"> <em>old_length1</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">int&nbsp;</td>
          <td class="paramname"> <em>old_length2</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">int&nbsp;</td>
          <td class="paramname"> <em>length1</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">int&nbsp;</td>
          <td class="paramname"> <em>length2</em></td><td>&nbsp;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td><td></td>
        </tr>
      </table>
</div>
<div class="memdoc">

<p>Enlargement of the dynamic programming matrix in case it is to small. </p>
<dl><dt><b>Parameters:</b></dt><dd>
  <table border="0" cellspacing="2" cellpadding="0">
    <tr><td valign="top"></td><td valign="top"><em>dyn_matrix</em>&nbsp;</td><td>The dynamic programming matrix. </td></tr>
    <tr><td valign="top"></td><td valign="top"><em>old_length1</em>&nbsp;</td><td>Current size of dimension 1. </td></tr>
    <tr><td valign="top"></td><td valign="top"><em>old_length2</em>&nbsp;</td><td>Current size of dimension 2. </td></tr>
    <tr><td valign="top"></td><td valign="top"><em>length1</em>&nbsp;</td><td>New size of dimension 1. </td></tr>
    <tr><td valign="top"></td><td valign="top"><em>length2</em>&nbsp;</td><td>New size of dimension 2. </td></tr>
  </table>
  </dd>
</dl>
<dl class="return"><dt><b>Returns:</b></dt><dd>Pointer to the new array. </dd></dl>

</div>
</div>
<a class="anchor" id="ad21db2c5bf21fdebaf4367bcec2165fb"></a><!-- doxytag: member="fastal.c::seq_pair2blast_diagonal" ref="ad21db2c5bf21fdebaf4367bcec2165fb" args="(char *seq_file_name1, char *seq_file_name2, int **diagonals, int *dig_length, int l1, int l2, int is_dna)" -->
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">int seq_pair2blast_diagonal </td>
          <td>(</td>
          <td class="paramtype">char *&nbsp;</td>
          <td class="paramname"> <em>seq_file_name1</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">char *&nbsp;</td>
          <td class="paramname"> <em>seq_file_name2</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">int **&nbsp;</td>
          <td class="paramname"> <em>diagonals</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">int *&nbsp;</td>
          <td class="paramname"> <em>dig_length</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">int&nbsp;</td>
          <td class="paramname"> <em>l1</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">int&nbsp;</td>
          <td class="paramname"> <em>l2</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">int&nbsp;</td>
          <td class="paramname"> <em>is_dna</em></td><td>&nbsp;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td><td></td>
        </tr>
      </table>
</div>
<div class="memdoc">

<p>Calculates the diagonals between two sequences. </p>
<p>Uses a k-tup index to choose diagonals. </p>
<dl><dt><b>Parameters:</b></dt><dd>
  <table border="0" cellspacing="2" cellpadding="0">
    <tr><td valign="top"></td><td valign="top"><em>seq_file1</em>&nbsp;</td><td>File with sequence 1. </td></tr>
    <tr><td valign="top"></td><td valign="top"><em>seq_file2</em>&nbsp;</td><td>File with sequence 2. </td></tr>
    <tr><td valign="top"></td><td valign="top"><em>diagonals</em>&nbsp;</td><td>An array where the diagonal points will be stored. </td></tr>
    <tr><td valign="top"></td><td valign="top"><em>dig_length</em>&nbsp;</td><td>length of <em>diagonals</em> . </td></tr>
    <tr><td valign="top"></td><td valign="top"><em>num_points</em>&nbsp;</td><td>Number of points in all diagonals. </td></tr>
  </table>
  </dd>
</dl>
<dl class="return"><dt><b>Returns:</b></dt><dd>number of diagonals; </dd></dl>

</div>
</div>
<a class="anchor" id="af19c3e4bd9a785b7904bd40249817880"></a><!-- doxytag: member="fastal.c::seq_pair2blastz_diagonal" ref="af19c3e4bd9a785b7904bd40249817880" args="(char *seq_file_name1, char *seq_file_name2, int **diagonals, int *dig_length, int l1, int l2, int is_dna)" -->
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">int seq_pair2blastz_diagonal </td>
          <td>(</td>
          <td class="paramtype">char *&nbsp;</td>
          <td class="paramname"> <em>seq_file_name1</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">char *&nbsp;</td>
          <td class="paramname"> <em>seq_file_name2</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">int **&nbsp;</td>
          <td class="paramname"> <em>diagonals</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">int *&nbsp;</td>
          <td class="paramname"> <em>dig_length</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">int&nbsp;</td>
          <td class="paramname"> <em>l1</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">int&nbsp;</td>
          <td class="paramname"> <em>l2</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">int&nbsp;</td>
          <td class="paramname"> <em>is_dna</em></td><td>&nbsp;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td><td></td>
        </tr>
      </table>
</div>
<div class="memdoc">

<p>Calculates the diagonals between two sequences. </p>
<p>Uses blastz to calculate the diagonals. </p>
<dl><dt><b>Parameters:</b></dt><dd>
  <table border="0" cellspacing="2" cellpadding="0">
    <tr><td valign="top"></td><td valign="top"><em>seq_file1</em>&nbsp;</td><td>File with sequence 1. </td></tr>
    <tr><td valign="top"></td><td valign="top"><em>seq_file2</em>&nbsp;</td><td>File with sequence 2. </td></tr>
    <tr><td valign="top"></td><td valign="top"><em>diagonals</em>&nbsp;</td><td>An array where the diagonal points will be stored. </td></tr>
    <tr><td valign="top"></td><td valign="top"><em>dig_length</em>&nbsp;</td><td>length of <em>diagonals</em> . </td></tr>
    <tr><td valign="top"></td><td valign="top"><em>num_points</em>&nbsp;</td><td>Number of points in all diagonals. </td></tr>
  </table>
  </dd>
</dl>
<dl class="return"><dt><b>Returns:</b></dt><dd>number of diagonals; </dd></dl>

</div>
</div>
<a class="anchor" id="adb7e15a5c6a9c998e952ec110ba8d45b"></a><!-- doxytag: member="fastal.c::seq_pair2diagonal_own" ref="adb7e15a5c6a9c998e952ec110ba8d45b" args="(char *seq1, char *seq2, int **diagonals, int *dig_length, int l1, int l2, int is_dna, int word_length)" -->
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">int seq_pair2diagonal_own </td>
          <td>(</td>
          <td class="paramtype">char *&nbsp;</td>
          <td class="paramname"> <em>seq1</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">char *&nbsp;</td>
          <td class="paramname"> <em>seq2</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">int **&nbsp;</td>
          <td class="paramname"> <em>diagonals</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">int *&nbsp;</td>
          <td class="paramname"> <em>dig_length</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">int&nbsp;</td>
          <td class="paramname"> <em>l1</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">int&nbsp;</td>
          <td class="paramname"> <em>l2</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">int&nbsp;</td>
          <td class="paramname"> <em>is_dna</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">int&nbsp;</td>
          <td class="paramname"> <em>word_length</em></td><td>&nbsp;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td><td></td>
        </tr>
      </table>
</div>
<div class="memdoc">

<p>Calculates the diagonals between two sequences. </p>
<p>Uses to calculate the diagonals. </p>
<dl><dt><b>Parameters:</b></dt><dd>
  <table border="0" cellspacing="2" cellpadding="0">
    <tr><td valign="top"></td><td valign="top"><em>seq_file1</em>&nbsp;</td><td>File with sequence 1. </td></tr>
    <tr><td valign="top"></td><td valign="top"><em>seq_file2</em>&nbsp;</td><td>File with sequence 2. </td></tr>
    <tr><td valign="top"></td><td valign="top"><em>diagonals</em>&nbsp;</td><td>An array where the diagonal points will be stored. </td></tr>
    <tr><td valign="top"></td><td valign="top"><em>dig_length</em>&nbsp;</td><td>length of <em>diagonals</em> . </td></tr>
    <tr><td valign="top"></td><td valign="top"><em>num_points</em>&nbsp;</td><td>Number of points in all diagonals. </td></tr>
  </table>
  </dd>
</dl>
<dl class="return"><dt><b>Returns:</b></dt><dd>number of diagonals; </dd></dl>

</div>
</div>
<a class="anchor" id="afbe437cab15b82544b8fc62221005018"></a><!-- doxytag: member="fastal.c::sparse_dyn" ref="afbe437cab15b82544b8fc62221005018" args="(Fastal_profile **profiles, Fastal_param *param_set, void *method_arguments_p, int is_dna, FILE *edit_file, FILE *prof_file, int number)" -->
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">int sparse_dyn </td>
          <td>(</td>
          <td class="paramtype"><a class="el" href="structFastal__profile.html">Fastal_profile</a> **&nbsp;</td>
          <td class="paramname"> <em>profiles</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype"><a class="el" href="structFastal__param.html">Fastal_param</a> *&nbsp;</td>
          <td class="paramname"> <em>param_set</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">void *&nbsp;</td>
          <td class="paramname"> <em>method_arguments_p</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">int&nbsp;</td>
          <td class="paramname"> <em>is_dna</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">FILE *&nbsp;</td>
          <td class="paramname"> <em>edit_file</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">FILE *&nbsp;</td>
          <td class="paramname"> <em>prof_file</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">int&nbsp;</td>
          <td class="paramname"> <em>number</em></td><td>&nbsp;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td><td></td>
        </tr>
      </table>
</div>
<div class="memdoc">

<p>One run of sparse dynamic programming. </p>
<dl><dt><b>Parameters:</b></dt><dd>
  <table border="0" cellspacing="2" cellpadding="0">
    <tr><td valign="top"></td><td valign="top"><em>profiles</em>&nbsp;</td><td>The profiles. </td></tr>
    <tr><td valign="top"></td><td valign="top"><em>param_set</em>&nbsp;</td><td>The fastal parameters. </td></tr>
    <tr><td valign="top"></td><td valign="top"><em>method_arguments_p</em>&nbsp;</td><td>The method arguments. </td></tr>
    <tr><td valign="top"></td><td valign="top"><em>is_dna</em>&nbsp;</td><td>Sequences are DNA (<em>is_dna</em> = 1) or protein. </td></tr>
    <tr><td valign="top"></td><td valign="top"><em>edit_file</em>&nbsp;</td><td>The edit file. </td></tr>
    <tr><td valign="top"></td><td valign="top"><em>prof_file</em>&nbsp;</td><td>the profile file. </td></tr>
    <tr><td valign="top"></td><td valign="top"><em>number</em>&nbsp;</td><td>Number of the parent node. </td></tr>
  </table>
  </dd>
</dl>
<dl class="return"><dt><b>Returns:</b></dt><dd>The length of the alignment. </dd></dl>

</div>
</div>
<a class="anchor" id="a72b967437b83b6ea91c93ff204415add"></a><!-- doxytag: member="fastal.c::sumup_profile" ref="a72b967437b83b6ea91c93ff204415add" args="(Fastal_profile *profile, int **sumup, Fastal_param *param_set)" -->
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">int** sumup_profile </td>
          <td>(</td>
          <td class="paramtype"><a class="el" href="structFastal__profile.html">Fastal_profile</a> *&nbsp;</td>
          <td class="paramname"> <em>profile</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">int **&nbsp;</td>
          <td class="paramname"> <em>sumup</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype"><a class="el" href="structFastal__param.html">Fastal_param</a> *&nbsp;</td>
          <td class="paramname"> <em>param_set</em></td><td>&nbsp;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td><td></td>
        </tr>
      </table>
</div>
<div class="memdoc">

<p>This method takes a profile and turns it into a sumed up version. </p>
<p>Required for NW-algorithm. </p>
<dl><dt><b>Parameters:</b></dt><dd>
  <table border="0" cellspacing="2" cellpadding="0">
    <tr><td valign="top"></td><td valign="top"><em>profile</em>&nbsp;</td><td>The profile to sum up. </td></tr>
    <tr><td valign="top"></td><td valign="top"><em>sumup</em>&nbsp;</td><td>A field where the result will be stored. </td></tr>
    <tr><td valign="top"></td><td valign="top"><em>param_set</em>&nbsp;</td><td>Parameters for the fastal algorithm. </td></tr>
  </table>
  </dd>
</dl>
<dl class="return"><dt><b>Returns:</b></dt><dd>The new <em>sumup</em>. </dd></dl>

</div>
</div>
<a class="anchor" id="a4f3eeb11dc9d50f60d94497ddff1421d"></a><!-- doxytag: member="fastal.c::write2file" ref="a4f3eeb11dc9d50f60d94497ddff1421d" args="(int **sumup_prf, int length, FILE *file, int number, int num_sequences, Fastal_param *param_set)" -->
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">void write2file </td>
          <td>(</td>
          <td class="paramtype">int **&nbsp;</td>
          <td class="paramname"> <em>sumup_prf</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">int&nbsp;</td>
          <td class="paramname"> <em>length</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">FILE *&nbsp;</td>
          <td class="paramname"> <em>file</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">int&nbsp;</td>
          <td class="paramname"> <em>number</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">int&nbsp;</td>
          <td class="paramname"> <em>num_sequences</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype"><a class="el" href="structFastal__param.html">Fastal_param</a> *&nbsp;</td>
          <td class="paramname"> <em>param_set</em></td><td>&nbsp;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td><td></td>
        </tr>
      </table>
</div>
<div class="memdoc">

<p>Writes a profile to a file. </p>
<dl><dt><b>Parameters:</b></dt><dd>
  <table border="0" cellspacing="2" cellpadding="0">
    <tr><td valign="top"></td><td valign="top"><em>sumup_prf</em>&nbsp;</td><td>The profile array, not a real profile. </td></tr>
    <tr><td valign="top"></td><td valign="top"><em>length</em>&nbsp;</td><td>The length of the profile. The format can be seen in ./test.txt </td></tr>
    <tr><td valign="top"></td><td valign="top"><em>file</em>&nbsp;</td><td>The FILE object to write the the profile into. </td></tr>
    <tr><td valign="top"></td><td valign="top"><em>is_dna</em>&nbsp;</td><td>The type of sequence. </td></tr>
    <tr><td valign="top"></td><td valign="top"><em>number</em>&nbsp;</td><td>The number of the profile. </td></tr>
  </table>
  </dd>
</dl>

</div>
</div>
</div>
<!--- window showing the filter options -->
<div id="MSearchSelectWindow"
     onmouseover="return searchBox.OnSearchSelectShow()"
     onmouseout="return searchBox.OnSearchSelectHide()"
     onkeydown="return searchBox.OnSearchSelectKey(event)">
<a class="SelectItem" href="javascript:void(0)" onclick="searchBox.OnSelectItem(0)"><span class="SelectionMark">&nbsp;</span>All</a><a class="SelectItem" href="javascript:void(0)" onclick="searchBox.OnSelectItem(1)"><span class="SelectionMark">&nbsp;</span>Classes</a><a class="SelectItem" href="javascript:void(0)" onclick="searchBox.OnSelectItem(2)"><span class="SelectionMark">&nbsp;</span>Files</a><a class="SelectItem" href="javascript:void(0)" onclick="searchBox.OnSelectItem(3)"><span class="SelectionMark">&nbsp;</span>Functions</a><a class="SelectItem" href="javascript:void(0)" onclick="searchBox.OnSelectItem(4)"><span class="SelectionMark">&nbsp;</span>Variables</a></div>

<!-- iframe showing the search results (closed by default) -->
<div id="MSearchResultsWindow">
<iframe src="" frameborder="0" 
        name="MSearchResults" id="MSearchResults">
</iframe>
</div>

<hr size="1"/><address style="text-align: right;"><small>Generated on 28 Sep 2012 for T-Coffee by&nbsp;
<a href="http://www.doxygen.org/index.html">
<img class="footer" src="doxygen.png" alt="doxygen"/></a> 1.6.1 </small></address>
</body>
</html>
