<!-- This comment will put IE 6, 7 and 8 in quirks mode -->
<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">
<html xmlns="http://www.w3.org/1999/xhtml">
<head>
<meta http-equiv="Content-Type" content="text/xhtml;charset=UTF-8"/>
<title>T-Coffee: t_coffee.c File Reference</title>
<link href="tabs.css" rel="stylesheet" type="text/css"/>
<link href="search/search.css" rel="stylesheet" type="text/css"/>
<script type="text/javaScript" src="search/search.js"></script>
<link href="doxygen.css" rel="stylesheet" type="text/css"/>
</head>
<body onload='searchBox.OnSelectItem(0);'>
<!-- Generated by Doxygen 1.6.1 -->
<script type="text/javascript"><!--
var searchBox = new SearchBox("searchBox", "search",false,'Search');
--></script>
<div class="navigation" id="top">
  <div class="tabs">
    <ul>
      <li><a href="index.html"><span>Main&nbsp;Page</span></a></li>
      <li><a href="pages.html"><span>Related&nbsp;Pages</span></a></li>
      <li><a href="annotated.html"><span>Classes</span></a></li>
      <li class="current"><a href="files.html"><span>Files</span></a></li>
      <li>
        <div id="MSearchBox" class="MSearchBoxInactive">
        <img id="MSearchSelect" src="search/search.png"
             onmouseover="return searchBox.OnSearchSelectShow()"
             onmouseout="return searchBox.OnSearchSelectHide()"
             alt=""/>
        <input type="text" id="MSearchField" value="Search" accesskey="S"
             onfocus="searchBox.OnSearchFieldFocus(true)" 
             onblur="searchBox.OnSearchFieldFocus(false)" 
             onkeyup="searchBox.OnSearchFieldChange(event)"/>
        <a id="MSearchClose" href="javascript:searchBox.CloseResultsWindow()"><img id="MSearchCloseImg" border="0" src="search/close.png" alt=""/></a>
        </div>
      </li>
    </ul>
  </div>
  <div class="tabs">
    <ul>
      <li><a href="files.html"><span>File&nbsp;List</span></a></li>
      <li><a href="globals.html"><span>File&nbsp;Members</span></a></li>
    </ul>
  </div>
</div>
<div class="contents">
<h1>t_coffee.c File Reference</h1>
<p>Main file of T-Coffee.  
<a href="#_details">More...</a></p>
<code>#include &lt;stdio.h&gt;</code><br/>
<code>#include &lt;stdlib.h&gt;</code><br/>
<code>#include &lt;math.h&gt;</code><br/>
<code>#include &lt;stdarg.h&gt;</code><br/>
<code>#include &lt;signal.h&gt;</code><br/>
<code>#include &lt;string.h&gt;</code><br/>
<code>#include &lt;unistd.h&gt;</code><br/>
<code>#include &lt;float.h&gt;</code><br/>
<code>#include &quot;io_lib_header.h&quot;</code><br/>
<code>#include &quot;util_lib_header.h&quot;</code><br/>
<code>#include &quot;dp_lib_header.h&quot;</code><br/>
<code>#include &quot;define_header.h&quot;</code><br/>
<code>#include &quot;t_coffee.h&quot;</code><br/>
<code>#include &quot;km_coffee_header.h&quot;</code><br/>
<code>#include &quot;debug.h&quot;</code><br/>
<table border="0" cellpadding="0" cellspacing="0">
<tr><td colspan="2"><h2>Functions</h2></td></tr>
<tr><td class="memItemLeft" align="right" valign="top">static int&nbsp;</td><td class="memItemRight" valign="bottom"><a class="el" href="t__coffee_8c.html#a8479a9cbdd5dd9fed34c1d09d483d7a7">run_other_pg</a> (int argc, char *argv[])</td></tr>
<tr><td class="mdescLeft">&nbsp;</td><td class="mdescRight">Use one of the other_pg options.  <a href="#a8479a9cbdd5dd9fed34c1d09d483d7a7"></a><br/></td></tr>
<tr><td class="memItemLeft" align="right" valign="top">int&nbsp;</td><td class="memItemRight" valign="bottom"><a class="el" href="t__coffee_8c.html#abb7d6a4d94006dc52ba2cf916c7bcc32">batch_main</a> (int argc, char **argv)</td></tr>
<tr><td class="mdescLeft">&nbsp;</td><td class="mdescRight">Brings everything together, but does not contain computations.  <a href="#abb7d6a4d94006dc52ba2cf916c7bcc32"></a><br/></td></tr>
<tr><td class="memItemLeft" align="right" valign="top"><a class="anchor" id="a0ddf1224851353fc92bfbff6f499fa97"></a><!-- doxytag: member="t_coffee.c::main" ref="a0ddf1224851353fc92bfbff6f499fa97" args="(int argc, char *argv[])" -->
int&nbsp;</td><td class="memItemRight" valign="bottom"><a class="el" href="t__coffee_8c.html#a0ddf1224851353fc92bfbff6f499fa97">main</a> (int argc, char *argv[])</td></tr>
<tr><td class="mdescLeft">&nbsp;</td><td class="mdescRight">Pass arguments to <a class="el" href="t__coffee_8c.html#abb7d6a4d94006dc52ba2cf916c7bcc32" title="Brings everything together, but does not contain computations.">batch_main</a>. <br/></td></tr>
</table>
<hr/><a name="_details"></a><h2>Detailed Description</h2>
<p>Main file of T-Coffee. </p>
<dl class="author"><dt><b>Author:</b></dt><dd>Cedric Notredame</dd></dl>
<hr/><h2>Function Documentation</h2>
<a class="anchor" id="abb7d6a4d94006dc52ba2cf916c7bcc32"></a><!-- doxytag: member="t_coffee.c::batch_main" ref="abb7d6a4d94006dc52ba2cf916c7bcc32" args="(int argc, char **argv)" -->
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">int batch_main </td>
          <td>(</td>
          <td class="paramtype">int&nbsp;</td>
          <td class="paramname"> <em>argc</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">char **&nbsp;</td>
          <td class="paramname"> <em>argv</em></td><td>&nbsp;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td><td></td>
        </tr>
      </table>
</div>
<div class="memdoc">

<p>Brings everything together, but does not contain computations. </p>
<h2><a class="anchor" id="remark">
Remark</a></h2>
<p>The follwowing text is just a summation of short lines of documentation appearing throughout the code of this function. Often they make more sense in their specific context, that means while reading the source code.</p>
<h2><a class="anchor" id="prep">
Preparation</a></h2>

<p><p>Before anything else, check if we want to use the <b>other_pg</b> option of T-Coffee. If so, redirect to <a class="el" href="t__coffee_8c.html#a8479a9cbdd5dd9fed34c1d09d483d7a7" title="Use one of the other_pg options.">run_other_pg</a>. (In case of kmcoffee, load the kmcoffee arguments from km_coffee)</p>
<p>Read all the parameters of T_Coffee using <a class="el" href="util_8c.html#a122df509fd59e72b03929c9536b4b521" title="Read command line parameters.">get_cl_param</a></p>
<p>Load T-Coffee default parameters. Next, decide whether a special mode of T_Coffee has been called and load the corresponing parameters from a function. Here's an example: </p>
<div class="fragment"><pre class="fragment"> <span class="keywordflow">else</span> <span class="keywordflow">if</span> ( strm (special_mode, <span class="stringliteral">&quot;psicoffee&quot;</span>)) new_arg = get_psicoffee_defaults( NULL,lseq_type );
</pre></div><p>Special commands, that usually end T-Coffee are handled next: check_configuation, update, version etc.</p>
<p>If T-Coffee is invoked without any arguments, it displays the available method names through display_method_names and exits.</p>
<p>Array <code>list_file</code>[] contains the names of all input files like sequences, profiles and templates, but also methods. There is a convention in T-Coffee to start the filename of with a capital letter specifying its type:</p>
<ul>
<li><b>M</b> for methods</li>
<li><b>R</b> for profiles</li>
<li><b>A</b> for alignments</li>
<li><b>L</b> for libraries (via the -lib flag)</li>
<li><b>P</b> for PDB structures</li>
</ul>
<p>See <a class="el" href="util__constraints__list_8c.html#a887d8f521ce7fd9abca25489de4fb159" title="Central method for reading different types of files.">read_constraint_list</a> for more information on how these files will be processed.</p>
<p>If no alignment A, library L or method M is in the list of infiles (that means in <code>list_files</code>[]), T-Coffee will add a default method here. The default method is currently <b>Mproba_pair</b>.</p>
<p>Check the content of each input file and report possible errors.</p>
<p>First non-error output: List the Input files, which are usually a sequence file (S) and one or several methods (M), as shown in the example: </p>
<div class="fragment"><pre class="fragment"> INPUT FILES
         Input File (S) sh3.fasta  Format fasta_seq
         Input File (M) proba_pair
</pre></div><h2><a class="anchor" id="input">
Input Sequences and Libraries</a></h2>
<p>Read the Sequences (<a class="el" href="util__constraints__list_8c.html#acaa0f2350cdd830d21cedacce4c937bd" title="Reads sequence files and alignments or can extract the sequence from a library.">read_seq_in_n_list</a>). Moreover, for these methods that need Blast scans on the sequences, precompute_blast_db is called. </p>
<dl class="note"><dt><b>Note:</b></dt><dd>The precompute_blast_db method is currently invoked only for methods that contain the substring "blast". As far as I know, it's main purpose is the method <b>blastr_pair</b> which is used in the special mode blastr. See the source code of precompute_blast_db for mmore information.</dd></dl>
<p>Translate Sequences from RNA to DNA or vice versa, if wanted (<a class="el" href="reformat_8c.html#abb50fda276ca9e708a96d152b331f9ab" title="Transforms RNA to DNA or the other way around.">transform_sequence</a>).</p>
<p>Reorder the sequences, if wanted, by using reorder_seq. You can make T-Coffee order the sequences lexicographically by their names by using the <code>-inorder=<code>aligned</code> flag</code>. If it is set, the order of the sequences is specified by sort_string_array(new_oder), otherwise the order from the input file (i.e. the order in the <a class="el" href="structSequence.html#a02358a7b87682bb92742be23051f2ba0" title="name[nseq ][MAXNAMELEN ] Names of the sequences">Sequence::name</a> array) is kept.</p>
<p>Prepare the global <a class="el" href="structConstraint__list.html" title="The central data structure (library) containing nearly all information to produce...">Constraint_list</a> object using declare_constraint_list. This function allocates memory for a <a class="el" href="structConstraint__list.html" title="The central data structure (library) containing nearly all information to produce...">Constraint_list</a> and sets the <a class="el" href="structConstraint__list.html#ab1d38cee44aa8748b481578fd0b280a4" title="Central structure containing the constraints/edges between residues.">Constraint_list::residue_index</a> to default values.</p>
<p>Identify Master Sequences for the master mode (option <code>-master</code>). Per default, master mode is not used (<code>-master=<code>no</code>)</code>, but there are several ways to specify master sequences. The master mode is experimental only, the idea is not to run all pairwise alignments, but only some of them. See prepare_master (best in the soruce code) for more details.</p>
<h2><a class="anchor" id="templates">
Get the Templates</a></h2>
<p>Call <a class="el" href="aln__convertion__util_8c.html#a4cc13549aa410debb81ff52c5d72635f" title="Adds templates to a Sequence object.">seq2template_seq</a> for each template file in <code>template_file_list</code>. This will read/fetch template files and incorporate them into the Seqeunce object. If no template file is given but the <a class="el" href="structSequence.html" title="Container for several sequences and their attributes.">Sequence</a> object already contains templates, these are written into a <code></code>.template_file file. Next,the same procedure is repeated for profile_template_files (<code>-profile_template_file</code>), using profile_seq2template_seq</p>
<dl class="see"><dt><b>See also:</b></dt><dd><a class="el" href="structTemplate.html" title="Any sort of Template like PDB structure, Profile or Secondary structure.">Template</a> for more information on how templates are stored within T-Coffee.</dd></dl>
<p>Add the sequences as their own template via the generic <a class="el" href="aln__convertion__util_8c.html#a4cc13549aa410debb81ff52c5d72635f" title="Adds templates to a Sequence object.">seq2template_seq</a> and the option <code>"SELF_S"</code>. Finally, specify the types of used templates in the <a class="el" href="structSequence.html" title="Container for several sequences and their attributes.">Sequence</a> object by <a class="el" href="aln__convertion__util_8c.html#a4ba6079dfdda4a2d7dfc5301eae9983c" title="Specify types of used templates for each sequence.">seq2template_type</a>.</p>
<p>Set further options like the potential substitution matrix, the <code>-filter_lib</code> flag, the profile_mode (see get_profile_mode_function), some BLAST parameters, some PDB-BLAST parameters and the multicore mode, before we can go on to the real stuff.</p>
<h2><a class="anchor" id="libcompilation">
Compile the Library</a></h2>
<h3><a class="anchor" id="read_CL">
Build Up the Initial Constraint_list</a></h3>
<p>Finally we can start with the actual algorithm to build an alignment. In this step we build the initial <a class="el" href="structConstraint__list.html" title="The central data structure (library) containing nearly all information to produce...">Constraint_list</a> from various sources. Usually, the initial <a class="el" href="structConstraint__list.html" title="The central data structure (library) containing nearly all information to produce...">Constraint_list</a> will contain the information from pairwise alignment methods (per default <code>proba_pair</code>), but of course there are different modes of T-Coffee that use structural pairwise alignments or the output of multiple aligners. However, this is the place where all this information is produced and collected. The function that summarizes all this is <a class="el" href="util__constraints__list_8c.html#a521d3d0c5498572f86fa9d0888b3f922" title="Read several types of input files and build the library.">read_n_constraint_list</a> and its documentation is recommended reading.</p>
<p>Just to give a short overview: It will distibute the process of generating pairwise alignments (i.e. calling external aligners or built-ins like proba_pair) over several CPUs and incorporate their outputs in one single <a class="el" href="structConstraint__list.html" title="The central data structure (library) containing nearly all information to produce...">Constraint_list</a> object called <b>CL</b>. There are some maybe non-intuitive decisions done in the process of reading the outputs, which should be documented in the function. </p>
<dl class="see"><dt><b>See also:</b></dt><dd><a class="el" href="structConstraint__list.html#ab1d38cee44aa8748b481578fd0b280a4" title="Central structure containing the constraints/edges between residues.">Constraint_list::residue_index</a> to understand how <b>CL</b> stores edges/constraints.</dd></dl>
<p>Some methods, like <code>proba_pair</code>, need a cleanup afterwards to free static memory allocated during their process. That's why free_pair_wise is called here.</p>
<p>If the <a class="el" href="structConstraint__list.html" title="The central data structure (library) containing nearly all information to produce...">Constraint_list</a> should, for some reason, be empty afterwards, report an Error.</p>
<p>If an <code>-extend</code> has been set to a value greater than 0 an initial filtering of the <a class="el" href="structConstraint__list.html" title="The central data structure (library) containing nearly all information to produce...">Constraint_list</a> is performed, see <a class="el" href="util__constraints__list_8c.html#a728a4f56bf44e8dd8e32d903d1ab2992" title="Deletes unimportant edges from the Constraint_list.">filter_constraint_list</a>.</p>
<p>Choose the extension mode given by the <code>-extend</code> flag, see <a class="el" href="evaluate_8c.html#af7729c6b2b25f72bc85faf7a79b26002" title="Set the extension mode.">choose_extension_mode</a>. Default values seems to be <code>very_fast_triplet</code> (??) </p>
<dl class="see"><dt><b>See also:</b></dt><dd><a class="el" href="structConstraint__list.html#ad277d230cff370e0bbe68d4057ebfc8a" title="Function for evaluating the cost of a pair of residues.">Constraint_list::evaluate_residue_pair</a> for more about the Extension mode.</dd></dl>
<h3><a class="anchor" id="extend_CL">
Consistency: Relaxation, Filtering, Extension!</a></h3>
<p>This is where the magic happens!</p>
<p>First, if <code>-out_lib</code> is used, the library will be written into the specified file via the save_constraint_list function. If you demand an extended library (by <code>-out_lib_mode</code> = <code>extended</code>) then first the extension steps are performed and then the library is written to a file via save_extended_constraint_list. If this was all you wanted to do (<code>lib_only</code>), exit the program afterwards.</p>
<p>Extension step:</p>
<ol type="1">
<li>If <code>filter_lib</code> &gt; 0 then reduce the <a class="el" href="structConstraint__list.html#ab1d38cee44aa8748b481578fd0b280a4" title="Central structure containing the constraints/edges between residues.">Constraint_list::residue_index</a> by removing edges with weight smaller or equal to <code>filter_lib</code>. Usually not used right here. See <a class="el" href="util__constraints__list_8c.html#a728a4f56bf44e8dd8e32d903d1ab2992" title="Deletes unimportant edges from the Constraint_list.">filter_constraint_list</a>.</li>
<li>Only if the environment variable <code>EXTEND4TC</code> is set to 1, call the <a class="el" href="util__constraints__list_8c.html#a4387a79f81d664dd6527ea412dec79e8" title="Apply Consistency.">extend_constraint_list</a>.</li>
<li>For a=0 to <code>relax_lib</code> (which is 1 per default) call <a class="el" href="util__constraints__list_8c.html#a588c769cc451ee98d557aef05e7f1664" title="Distributes the relaxation, see fork_relax_constraint_list.">relax_constraint_list</a>.</li>
<li>For a=0 to <code>shrink_lib</code> call shrink_constraint_list, but <code>schrink_lib</code> is usually set to 0.</li>
<li>A call to <a class="el" href="util__constraints__list_8c.html#a4789610f29c2539926ec98a67140223e" title="Determine maximal constraints for normalisation.">evaluate_constraint_list_reference</a> determines the maximum edge weights.</li>
</ol>
<p>We recommend reading the documentation of each of these functions to get to grips with the extension.</p>
<p>If <code>weight</code> is not set to no (but to <code>sim</code> or <code>matrix</code> for example), then compute a new distance matrix (cl2distance_matrix) and <a class="el" href="util__constraints__list_8c.html#a01c8291a43effd369ba50210a9863221" title="Specify a weight for each sequence in the Constraint_list.">weight_constraint_list</a> with the <code>seq_weight</code> flag.</p>
<p>Quadruplet mode (<b>experimental</b>). Only if <code>nseq_for_quadruplet</code> &gt; 0.</p>
<h2><a class="anchor" id="prog_aln">
Progressive Alignment</a></h2>
<p>At first, transform the Seqeunce into an Alignment via the small function <a class="el" href="aln__convertion__util_8c.html#a71ac4c9e04c3ccb5b8f2796f8e59343e" title="Transform a Sequence into an Alignment.">seq2aln</a> and use ungap_array to remove unwanted gaps. Then you have to choose the right <code>msa_mode</code> for evaluating columns. Right now, there are several options in the code, but only a few of them are used in general.</p>
<ul>
<li><code>sorted_aln</code> </li>
<li><code>seq_aln</code> </li>
</ul>
<dl class="todo"><dt><b><a class="el" href="todo.html#_todo000002">Todo:</a></b></dt><dd>The documentation of this function is not yet finished </dd></dl>
<p>Done. </p>
</p>

<p><div class="dynheader">
Here is the call graph for this function:</div>
<div class="dynsection">
<div class="center"><img src="t__coffee_8c_abb7d6a4d94006dc52ba2cf916c7bcc32_cgraph.png" border="0" usemap="#t__coffee_8c_abb7d6a4d94006dc52ba2cf916c7bcc32_cgraph_map" alt=""></div>
<map name="t__coffee_8c_abb7d6a4d94006dc52ba2cf916c7bcc32_cgraph_map" id="t__coffee_8c_abb7d6a4d94006dc52ba2cf916c7bcc32_cgraph">
<area shape="rect" id="node3" href="evaluate_8c.html#af7729c6b2b25f72bc85faf7a79b26002" title="Set the extension mode." alt="" coords="189,5,379,35"/><area shape="rect" id="node5" href="util__constraints__list_8c.html#a4789610f29c2539926ec98a67140223e" title="Determine maximal constraints for normalisation." alt="" coords="155,59,413,88"/><area shape="rect" id="node7" href="util__constraints__list_8c.html#a4387a79f81d664dd6527ea412dec79e8" title="Apply Consistency." alt="" coords="197,112,371,141"/><area shape="rect" id="node9" href="util__constraints__list_8c.html#a728a4f56bf44e8dd8e32d903d1ab2992" title="Deletes unimportant edges from the Constraint_list." alt="" coords="205,165,363,195"/><area shape="rect" id="node11" href="util_8c.html#a122df509fd59e72b03929c9536b4b521" title="Read command line parameters." alt="" coords="228,219,340,248"/><area shape="rect" id="node13" href="util__constraints__list_8c.html#a521d3d0c5498572f86fa9d0888b3f922" title="Read several types of input files and build the library." alt="" coords="199,272,369,301"/><area shape="rect" id="node15" href="util__constraints__list_8c.html#acaa0f2350cdd830d21cedacce4c937bd" title="Reads sequence files and alignments or can extract the sequence from a library." alt="" coords="213,325,355,355"/><area shape="rect" id="node17" href="util__constraints__list_8c.html#a588c769cc451ee98d557aef05e7f1664" title="Distributes the relaxation, see fork_relax_constraint_list." alt="" coords="205,379,363,408"/><area shape="rect" id="node19" href="t__coffee_8c.html#a8479a9cbdd5dd9fed34c1d09d483d7a7" title="Use one of the other_pg options." alt="" coords="229,432,339,461"/><area shape="rect" id="node23" href="aln__convertion__util_8c.html#a71ac4c9e04c3ccb5b8f2796f8e59343e" title="Transform a Sequence into an Alignment." alt="" coords="248,485,320,515"/><area shape="rect" id="node25" href="aln__convertion__util_8c.html#a4cc13549aa410debb81ff52c5d72635f" title="Adds templates to a Sequence object." alt="" coords="209,539,359,568"/><area shape="rect" id="node27" href="aln__convertion__util_8c.html#a4ba6079dfdda4a2d7dfc5301eae9983c" title="Specify types of used templates for each sequence." alt="" coords="205,592,363,621"/><area shape="rect" id="node29" href="reformat_8c.html#abb50fda276ca9e708a96d152b331f9ab" title="Transforms RNA to DNA or the other way around." alt="" coords="204,645,364,675"/><area shape="rect" id="node31" href="util__constraints__list_8c.html#a01c8291a43effd369ba50210a9863221" title="Specify a weight for each sequence in the Constraint_list." alt="" coords="199,699,369,728"/><area shape="rect" id="node21" href="fastal_8c.html#aa9c66bc55318e62544273983419ee51f" title="main of the fastal algorithm" alt="" coords="461,432,560,461"/></map>
</div>
</p>

</div>
</div>
<a class="anchor" id="a8479a9cbdd5dd9fed34c1d09d483d7a7"></a><!-- doxytag: member="t_coffee.c::run_other_pg" ref="a8479a9cbdd5dd9fed34c1d09d483d7a7" args="(int argc, char *argv[])" -->
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">int run_other_pg </td>
          <td>(</td>
          <td class="paramtype">int&nbsp;</td>
          <td class="paramname"> <em>argc</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">char *&nbsp;</td>
          <td class="paramname"> <em>argv</em>[]</td><td>&nbsp;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td><td><code> [static]</code></td>
        </tr>
      </table>
</div>
<div class="memdoc">

<p>Use one of the other_pg options. </p>
<p>If T-Coffee is called with the <b>other_pg</b> option, this function determines what to do.</p>
<dl class="see"><dt><b>See also:</b></dt><dd>seq_reformat, aln_compare </dd></dl>

<p><div class="dynheader">
Here is the call graph for this function:</div>
<div class="dynsection">
<div class="center"><img src="t__coffee_8c_a8479a9cbdd5dd9fed34c1d09d483d7a7_cgraph.png" border="0" usemap="#t__coffee_8c_a8479a9cbdd5dd9fed34c1d09d483d7a7_cgraph_map" alt=""></div>
<map name="t__coffee_8c_a8479a9cbdd5dd9fed34c1d09d483d7a7_cgraph_map" id="t__coffee_8c_a8479a9cbdd5dd9fed34c1d09d483d7a7_cgraph">
<area shape="rect" id="node3" href="fastal_8c.html#aa9c66bc55318e62544273983419ee51f" title="main of the fastal algorithm" alt="" coords="165,5,264,35"/></map>
</div>
</p>

</div>
</div>
</div>
<!--- window showing the filter options -->
<div id="MSearchSelectWindow"
     onmouseover="return searchBox.OnSearchSelectShow()"
     onmouseout="return searchBox.OnSearchSelectHide()"
     onkeydown="return searchBox.OnSearchSelectKey(event)">
<a class="SelectItem" href="javascript:void(0)" onclick="searchBox.OnSelectItem(0)"><span class="SelectionMark">&nbsp;</span>All</a><a class="SelectItem" href="javascript:void(0)" onclick="searchBox.OnSelectItem(1)"><span class="SelectionMark">&nbsp;</span>Classes</a><a class="SelectItem" href="javascript:void(0)" onclick="searchBox.OnSelectItem(2)"><span class="SelectionMark">&nbsp;</span>Files</a><a class="SelectItem" href="javascript:void(0)" onclick="searchBox.OnSelectItem(3)"><span class="SelectionMark">&nbsp;</span>Functions</a><a class="SelectItem" href="javascript:void(0)" onclick="searchBox.OnSelectItem(4)"><span class="SelectionMark">&nbsp;</span>Variables</a></div>

<!-- iframe showing the search results (closed by default) -->
<div id="MSearchResultsWindow">
<iframe src="" frameborder="0" 
        name="MSearchResults" id="MSearchResults">
</iframe>
</div>

<hr size="1"/><address style="text-align: right;"><small>Generated on 28 Sep 2012 for T-Coffee by&nbsp;
<a href="http://www.doxygen.org/index.html">
<img class="footer" src="doxygen.png" alt="doxygen"/></a> 1.6.1 </small></address>
</body>
</html>
